File: benchmarks/swift/b12_ga.swift
// Genetic Algorithm minimizing Rosenbrock — Computational Intelligence
// benchmark (MICAI). Tournament selection, uniform crossover/mutation; all
// +,-,*,/ so bit-exact. Shared LCG (seed 42) drives every random decision.
import Foundation
let D = 30, P = 5000, G = 1200, T = 3
let MUT_RATE = 0.1, MUT_STEP = 0.1
var state: UInt64 = 42
@inline(__always) func nextDouble() -> Double {
    state = state &* 6364136223846793005 &+ 1442695040888963407
    return Double(state >> 33) / Double(UInt64(1) << 31)
}
func rosenbrock(_ x: [Double], _ off: Int) -> Double {
    var f = 0.0
    for i in 0..<(D - 1) { let a = x[off+i+1] - x[off+i] * x[off+i]; let b = 1.0 - x[off+i]; f += 100.0 * a * a + b * b }
    return f
}
var pop = [Double](repeating: 0, count: P * D)
var newpop = [Double](repeating: 0, count: P * D)
var fitness = [Double](repeating: 0, count: P)
var bestGenes = [Double](repeating: 0, count: D)
for p in 0..<P { for d in 0..<D { pop[p*D+d] = (nextDouble() * 2 - 1) * 5.0 } }
var bestFit = 1e300
for _ in 0..<G {
    for p in 0..<P {
        let f = rosenbrock(pop, p*D); fitness[p] = f
        if f < bestFit { bestFit = f; for d in 0..<D { bestGenes[d] = pop[p*D+d] } }
    }
    for d in 0..<D { newpop[0*D+d] = bestGenes[d] }
    for i in 1..<P {
        var a = Int(nextDouble() * Double(P))
        for _ in 1..<T { let idx = Int(nextDouble() * Double(P)); if fitness[idx] < fitness[a] { a = idx } }
        var b = Int(nextDouble() * Double(P))
        for _ in 1..<T { let idx = Int(nextDouble() * Double(P)); if fitness[idx] < fitness[b] { b = idx } }
        for d in 0..<D { newpop[i*D+d] = (nextDouble() < 0.5) ? pop[a*D+d] : pop[b*D+d] }
        for d in 0..<D { if nextDouble() < MUT_RATE { newpop[i*D+d] += (nextDouble() * 2 - 1) * MUT_STEP } }
    }
    for p in 0..<P { for d in 0..<D { pop[p*D+d] = newpop[p*D+d] } }
}
var gsum = 0.0
for d in 0..<D { gsum += bestGenes[d] }
print(String(format: "%.6f %.6f", bestFit, gsum))

File: benchmarks/zig/b12_ga.zig
// Genetic Algorithm minimizing Rosenbrock — Computational Intelligence
// benchmark (MICAI). Tournament selection, uniform crossover/mutation; all
// +,-,*,/ so bit-exact. Shared LCG (seed 42) drives every random decision.
const std = @import("std");
const D = 30;
const P = 5000;
const G = 1200;
const T = 3;
const MUT_RATE = 0.1;
const MUT_STEP = 0.1;
var state: u64 = 42;
fn nextDouble() f64 {
    state = state *% 6364136223846793005 +% 1442695040888963407;
    return @as(f64, @floatFromInt(state >> 33)) / @as(f64, @floatFromInt(@as(u64, 1) << 31));
}
var pop: [P * D]f64 = undefined;
var newpop: [P * D]f64 = undefined;
var fitness: [P]f64 = undefined;
var best_genes: [D]f64 = undefined;
fn rosenbrock(off: usize) f64 {
    var f: f64 = 0.0;
    for (0..D - 1) |i| {
        const a = pop[off + i + 1] - pop[off + i] * pop[off + i];
        const b = 1.0 - pop[off + i];
        f += 100.0 * a * a + b * b;
    }
    return f;
}
pub fn main(init: std.process.Init) !void {
    const io = init.io;
    for (0..P) |p| for (0..D) |d| {
        pop[p * D + d] = (nextDouble() * 2.0 - 1.0) * 5.0;
    };
    var best_fit: f64 = 1e300;
    for (0..G) |_| {
        for (0..P) |p| {
            const f = rosenbrock(p * D);
            fitness[p] = f;
            if (f < best_fit) {
                best_fit = f;
                for (0..D) |d| {
                    best_genes[d] = pop[p * D + d];
                }
            }
        }
        for (0..D) |d| {
            newpop[d] = best_genes[d];
        }
        for (1..P) |i| {
            var a: usize = @intFromFloat(nextDouble() * P);
            for (1..T) |_| {
                const idx: usize = @intFromFloat(nextDouble() * P);
                if (fitness[idx] < fitness[a]) a = idx;
            }
            var b: usize = @intFromFloat(nextDouble() * P);
            for (1..T) |_| {
                const idx: usize = @intFromFloat(nextDouble() * P);
                if (fitness[idx] < fitness[b]) b = idx;
            }
            for (0..D) |d| {
                newpop[i * D + d] = if (nextDouble() < 0.5) pop[a * D + d] else pop[b * D + d];
            }
            for (0..D) |d| {
                if (nextDouble() < MUT_RATE)
                    newpop[i * D + d] += (nextDouble() * 2.0 - 1.0) * MUT_STEP;
            }
        }
        for (0..P) |p| for (0..D) |d| {
            pop[p * D + d] = newpop[p * D + d];
        };
    }
    var gsum: f64 = 0.0;
    for (0..D) |d| {
        gsum += best_genes[d];
    }
    var stdout_buffer: [256]u8 = undefined;
    var stdout_writer = std.Io.File.stdout().writer(io, &stdout_buffer);
    const stdout = &stdout_writer.interface;
    try stdout.print("{d:.6} {d:.6}\n", .{ best_fit, gsum });
    try stdout.flush();
}
